### DISCUSSION TITLE
`vips_linear` to invert image

### DISCUSSION THREAD

**Discussion by Jacherr** [libvips/libvips#3785] [Q&A]
Hi,

I'm attempting to invert an image while maintaining the alpha channel of the image. After a bit of searching I ran into [this](https://github.com/libvips/libvips/issues/1711#issuecomment-656553818), which seemed perfect for what I was doing.

The equivalent C, as far I can tell, should be 
```c
        // create image
	int a[4] = { -1, -1, -1, 1 };
	int b[4] = { 255, 255, 255, 0 };
	vips_linear(image, &image, a, b, 4, NULL);
```

However, when attempting to do this, the output is extremely artifacted and not in any resembling of what the command-line equivalent does.

For example, this:

![input](https://github.com/libvips/libvips/assets/40544507/ad9db8f5-5fef-49e5-8c30-ef62b7f15fb3)

Becomes this:

![real output](https://github.com/libvips/libvips/assets/40544507/4199f451-1511-400b-8f53-a9078649baf8)

I've been poking at this for a bit and might be missing something obvious, any ideas? 

Edit: Converting the input arrays to `double[]` doesn't fix this, but instead just produces a different garbled output.

**Comment by Jacherr** [comment]
Upon further investigation it would appear that the issues were caused by converting the image format to a double through this function. Casting it back to `uchar` and then processing it further fixed these issues.

**Reply by jcupitt** [reply]
Hi @Jacherr,

You can also pass the `uchar` flag to linear to get an 8 bit result, eg.:

```C
	vips_linear(image, &image, a, b, 4, "uchar", TRUE, NULL);
```
